examples/cosmos/ics20/denomTrace.qnt

```diff
@@ -67,6 +67,20 @@ module denomTrace {
       { path: [hopDestination].concat(denomTrace.path), baseDenom: denomTrace.baseDenom }
     }
   }
+
+  pure def isProperChannelSetup(setup: str -> str -> str): bool =
+    pure val chainNames = Set("A", "B", "C") // should be channelSetup.keys(), but that's breaking the effect checker somehow (see issue #808)
+    and {
+      // Check the domain for the second level of chains
+      chainNames.forall(chain => setup.get(chain).keys() == chainNames),
+      tuples(chainNames, chainNames).forall((fromChain, toChain) =>
+        fromChain != toChain implies // we dont care about channels from chain to itself
```

Comment:
I couldn't think of a way to define the set of all possible channel setups without having channels from chains to themselves in it. If you can think of a way, please let me know. My solution was to ignore them in here, but it's annoying that they appear in the output (i.e. `Map("A" -> Map("A" -> "thing-I-dont-care-about" ...) ...)`

Reply:
In the synopsis of the [connection](https://github.com/cosmos/ibc/tree/main/spec/core/ics-003-connection-semantics) spec it says a connection is between two separate chains. As channels are linked to connections, I guess self-loops are ruled out here. Although I didn't find the check in the connection spec. But it might make sense to dig there.

Reply:
[ICS 4](https://github.com/cosmos/ibc/blob/8d9d3b6fe7309b034df8457760b3bbd11d24b8e1/spec/core/ics-004-channel-and-packet-semantics/README.md?plain=1#L42) (bold emphasis is mine):

> A *channel* is a pipeline for exactly-once packet delivery between specific modules on **separate blockchains**, which has at least one end capable of sending packets and one end capable of receiving packets.